### PR TITLE
Adding the platform finding feature.

### PR DESCRIPTION
**Pull request by ticcky**
- Adding platform finding feature.
 - Removing useless DM's from the codebase (UfalRuleDM, PUfalRuleDM).
 - Making the policy code more readable.
 - Improving 'last-talked-about' resolution so that platform functionality and the old direction finding work well together.
 - Adding 'remove' method to ConfusionNetwork to implement the previous point.
 - Fixing printing unicode logs to non-unicode terminal.

### CODE REVIEW

File: alex/applications/thub.py [hunk 1]
```diff
@@ -213,10 +220,11 @@ def run(self):
 
         except KeyboardInterrupt:
             print 'KeyboardInterrupt exception in: %s' % multiprocessing.current_process().name
-            self.close_event.set()
+            #self.close_event.set()
             return
         except:
             self.cfg['Logging']['system_logger'].exception('Uncaught exception in THUB process.')
+            #self.close_event.set()
```
Comment (oplatek) [inline alex/applications/thub.py]: Seems little suspicious for me as well as 4 rows above

File: alex/applications/PublicTransportInfoCS/data/train_names.expanded.txt [hunk 2]
```diff
@@ -0,0 +1,178 @@
+albert einstein	albert einstein
```
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/data/train_names.expanded.txt]: What's the difference between "expanded" and "plain" train names? Are these supposed to be inflected in all cases?

File: alex/applications/PublicTransportInfoCS/directions.py [hunk 3]
```diff
@@ -482,9 +486,117 @@ def search_stop(self, stop_mask, city=None, max_count=0, skip_count=0):
             TTLANG.ENGLISH # language
         )
 
+    def search_train_station(self, stop_mask, max_count=0, skip_count=0):
```
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/directions.py]: Is this here because we don't know which stations are for trains from our database?

File: alex/applications/PublicTransportInfoCS/directions.py [hunk 4]
```diff
@@ -482,9 +486,117 @@ def search_stop(self, stop_mask, city=None, max_count=0, skip_count=0):
...
+
+
+    @lru_cache(maxsize=10)
+    def get_platform(self, platform_info):
+        # try to map from-to to IDOS identifiers, default to originals
+        self.system_logger.info("ALEX: Looking up platform for: %s -- %s" %
+                                (platform_info.from_stop,
+                                 platform_info.to_stop))
+
+        if platform_info.from_city != 'none' and platform_info.from_stop != \
```
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/directions.py]: I'd rather use a bit longer lines here.

File: alex/applications/PublicTransportInfoCS/hdc_policy.py [hunk 5]
```diff
@@ -175,16 +176,39 @@ def get_da(self, dialogue_state):
         # reset all slots depending on changed slots
         self.reset_on_change(dialogue_state, changed_slots)
 
+        c = {
```
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/hdc_policy.py]: What's "c"?

File: alex/applications/PublicTransportInfoCS/nlg_templates.cfg [hunk 6]
```diff
@@ -913,4 +913,24 @@ templates = {
...
     'apology()&inform(in_city={in_city})':
         "Omlouvám se, počasí pro obec {in_city} nemám k dispozici.",
+
+    #
+    # Platform info related templates.
+    #
+    'request(to_stop)&request(train_name)':
+        "A jakým směrem chcete jet? Nebo jak se jmenuje vlak?",
+    'inform(platform=none)&inform(track=none)&inform(direction={direction})':
+        "Ještě nění známo odkud to směrem {direction} jede.",
```
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/nlg_templates.cfg]: This word order sounds a bit unnatural to me – I'd use "odkud to jede směrem {...}".

File: alex/applications/PublicTransportInfoCS/nlg_templates.cfg [hunk 7]
```diff
@@ -913,4 +913,24 @@ templates = {
 
     'apology()&inform(in_city={in_city})':
         "Omlouvám se, počasí pro obec {in_city} nemám k dispozici.",
+
+    #
+    # Platform info related templates.
+    #
+    'request(to_stop)&request(train_name)':
+        "A jakým směrem chcete jet? Nebo jak se jmenuje vlak?",
```
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/nlg_templates.cfg]: How about "Jak se jmenuje váš vlak?"

File: alex/applications/PublicTransportInfoCS/platform_info.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+    def __unicode__(self):
+        return u"PlatformFinderResult(platform=%s, track=%s, direction=%s)" % (
+            unicode(self.platform),
+            unicode(self.track),
+            unicode(self.direction), )
+
+
+class CRWSPlatformInfo(object):
+    station_name_splitter = re.compile(r'[^a-zěščřžýáíéďťňóůú]*', re.UNICODE)
```
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/platform_info.py]: How about just \W ?

**Comment by tuetschek** [comment]
Just a side note – it would be nicer if you divided this into more pull requests (just platform-related stuff, just removing useless DM's etc.)

File: alex/applications/PublicTransportInfoCS/data/train_names.expanded.txt (hunk 2 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/data/train_names.expanded.txt]: It should be the same as for the stops - so yes in reality it should be probably all of the inflected forms. I will need to fix this along with expanding the list of train names in a next pull request. The train name feature is not complete yet (don't know where to reliably take the train names).

File: alex/applications/PublicTransportInfoCS/hdc_policy.py (hunk 5 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/hdc_policy.py]: It stands for 'condition'. I didn't want a long identifier. But maybe 'fact' or 'prop' (proposition) would be more intuitive?

File: alex/applications/PublicTransportInfoCS/nlg_templates.cfg (hunk 7 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/nlg_templates.cfg]: Fixed.

File: alex/applications/PublicTransportInfoCS/nlg_templates.cfg (hunk 6 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/nlg_templates.cfg]: Fixed. Thanks!

File: alex/applications/PublicTransportInfoCS/platform_info.py (hunk 8 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/platform_info.py]: That would be locale-dependent. But maybe I should add the list of characters somewhere to PTICS.utils?

File: alex/applications/thub.py (hunk 1 above)
Comment (ticcky) [inline alex/applications/thub.py]: Which 4 rows? This is a mistake introduced in https://github.com/UFAL-DSG/alex/commit/e97f1fc7938ad25f1681e27149533ea953a7a583 I will delete the lines.

File: alex/applications/PublicTransportInfoCS/platform_info.py (hunk 8 above)
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/platform_info.py]: If I understand correctly the paragraph for `\W` at https://docs.python.org/2/library/re.html, it won't be locale-dependent unless you set `re.LOCALE` .

File: alex/applications/PublicTransportInfoCS/hdc_policy.py (hunk 5 above)
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/hdc_policy.py]: My vote for `fact`. And a an explaining comment.

File: alex/applications/PublicTransportInfoCS/data/train_names.expanded.txt (hunk 2 above)
Comment (tuetschek) [inline alex/applications/PublicTransportInfoCS/data/train_names.expanded.txt]: Shouldn't the Czech Rail provide a list somewhere? Anyway, I wonder whether we should add this to the SVN, or just offer it as a download (same as stops, cities etc.).

File: alex/applications/PublicTransportInfoCS/platform_info.py (hunk 8 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/platform_info.py]: Fixed.

File: alex/applications/PublicTransportInfoCS/data/train_names.expanded.txt (hunk 2 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/data/train_names.expanded.txt]: We agreed we will keep the list in the base form in the repository. Not the expanded version.

File: alex/applications/PublicTransportInfoCS/directions.py (hunk 3 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/directions.py]: I added a comment.

File: alex/applications/PublicTransportInfoCS/directions.py (hunk 4 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/directions.py]: Fixed.

File: alex/applications/PublicTransportInfoCS/hdc_policy.py (hunk 5 above)
Comment (ticcky) [inline alex/applications/PublicTransportInfoCS/hdc_policy.py]: Fixed.

File: alex/components/dm/base.py [hunk 9]
```diff
@@ -214,14 +208,3 @@ def log_state(self):
         :return: none
         """
         self.dialogue_state.log_state()
-
-    def get_token(self):
```
Comment (jurcicek) [inline alex/components/dm/base.py]: Why was this removed? Where is the replacement?

File: alex/components/dm/ontology.py [hunk 10]
```diff
@@ -66,25 +66,27 @@ def slots_system_selects(self):
         return [slot for slot in self.ontology['slots'] if 'system_selects' in self.ontology['slot_attributes'][slot]]
 
     @lru_cache(maxsize=1000)
-    def last_talked_about(self, dat, name, value):
-        """Returns a list of slots and values that should be used to for tracking about what was talked about recently,
+    def last_talked_about(self, da_type, name, value):
+        """Returns a list of slots and values that should be used to for
```
Comment (jurcicek) [inline alex/components/dm/ontology.py]: Can you remove reformat the comments so that full 120 characters per line are used?

File: alex/components/dm/ontology.py [hunk 11]
```diff
@@ -66,25 +66,27 @@ def slots_system_selects(self):
...
+        tracking about what was talked about recently,
         given the input dialogue acts.
 
-        :param dat: the source dialogue act type
+        :param da_type: the source dialogue act type
         :param name: the source slot name
         :param value: the source slot value
-        :return: returns a list of target slot names and values used for tracking
+        :return: returns a list of target slot names and values used for
```
Comment (jurcicek) [inline alex/components/dm/ontology.py]: This line does not need to be split in two.

File: alex/components/dm/base.py (hunk 9 above)
Comment (ticcky) [inline alex/components/dm/base.py]: I didn't find it was being used anywhere, and I think I introduced it some time ago, so I thought I'd clean it up.

File: alex/components/dm/ontology.py (hunk 10 above)
Comment (ticcky) [inline alex/components/dm/ontology.py]: Fixed.

File: alex/components/dm/ontology.py (hunk 11 above)
Comment (ticcky) [inline alex/components/dm/ontology.py]: Fixed.

File: alex/components/dm/base.py (hunk 9 above)
Comment (jurcicek) [inline alex/components/dm/base.py]: It is being used by Martin Vejman I believe.

Comment (ticcky) [inline alex/components/dm/base.py]: OK. I moved it to utils/token.py because I think this is a better place for such method as it is not a general aspect of a dialog manager.

Comment (jurcicek) [inline alex/components/dm/base.py]: OK!

Best regards,
Filip Jurcicek

-------------------------------------------------------
Work tel. (CZ): +420221914402
Personal tel. (CZ): +420777805048
Skype: bozskyfilip

http://ufal.mff.cuni.cz/filip-jurcicek
http://ufal.mff.cuni.cz/grants

**Comment by ticcky** [comment]
I fixed some landscape.io problems that were relevant to my changes.

**Comment by jurcicek** [comment]
I am testing the platform task and it needs more work.

1st) to fix NLG templates:

It generates

Do zastávky Smíchovské nádraží. Spoj do Smíchovské nádraží nebyl nalezen.

Spoj do Smíchovské nádraží  should rather be "Spoj ve smeru Smi..."

I believe that there are more templates like this.

2nd)

switching between find_platform and find_connection is not ideal.


U: nastupiště

task = find_platform

S: Jaká je Vaše výchozí zastávka?

U: PRAHA HLAVNÍ NÁDRAŽÍ

S: Ze zastávky Praha, Hlavní nádraží. A jakým směrem chcete jet? Nebo jak
se jmenuje váš vlak?

U: CHTĚL BYCH JET DO BRNA

task = find_connection


Co-existence of both tasks: find_platform , find_connection should be
improved.


Best regards,
Filip Jurcicek

-------------------------------------------------------
Work tel. (CZ): +420221914402
Personal tel. (CZ): +420777805048
Skype: bozskyfilip

http://ufal.mff.cuni.cz/filip-jurcicek
http://ufal.mff.cuni.cz/grants
